File: ingestion/nodejs_fallback/src/engine/Runner.ts
// =============================================================================
// File: engine/Runner.ts
// Purpose: Generic EtLT engine — mirrors Python's engine/runner.py.
//   Reads pipeline_manifest.json, resolves dependency order, and executes
//   the Extract → stage-raw → Transform → stage-processed → Load flow
//   for each enabled source.
//
//   Each source entry in the manifest declares:
//     sheet_id / sheet_name  – where to extract data from
//     raw_blob               – GCS path for the raw staged CSV
//     processed_blob         – GCS path for the cleaned CSV (null if no cleaner)
//     cleaner.enabled        – whether to run a transform step
//     cleaner.path           – dotted module path, e.g. "EtLT.transform.expenses"
//                              module must export:  clean(rows) -> rows
//     loader.enabled         – whether to load into MySQL
//     loader.path            – dotted module path, e.g. "EtLT.load.transactions"
//                              module must export:  load() -> Promise<number>
//     depends_on             – list of source ids that must run first
// =============================================================================

import * as fs from "fs";
import * as path from "path";
import Papa from "papaparse";

import { appendCsvToGCS, downloadCsvFromGCS, uploadCsvToGCS } from "../EtLT/staging/GcsService";
import { fetchSheetData, parseOjectFromSheet, parseCsvFromObject } from "../EtLT/extract/SpreadsheetService";

const MANIFEST_PATH = process.env["MANIFEST_PATH"] ??
  path.resolve(__dirname, "../../../pipeline_manifest.json");

// ─── Types ────────────────────────────────────────────────────────────────────

interface CleanerConfig { enabled: boolean; path: string | null; }
interface LoaderConfig  { enabled: boolean; path: string; }

interface Source {
  id:             string;
  enabled:        boolean;
  sheet_id:       string;
  sheet_name:     string;
  raw_blob:       string;
  processed_blob: string | null;
  cleaner:        CleanerConfig;
  loader:         LoaderConfig;
  depends_on:     string[];
}

export interface SourceResult {
  status:         "ok" | "error";
  rows_in_sheet?: number;
  gcs_raw?:       string;
  gcs_processed?: string;
  rows_loaded?:   number;
  error?:         string;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/** Load a module by dotted path (e.g. "EtLT.load.plans" → src/EtLT/load/plans). */
function importModule(dottedPath: string): any {
  const parts = dottedPath.split(".");
  const absolutePath = path.resolve(__dirname, "..", ...parts);
  return require(absolutePath);
}

/** Topological sort — ensures every dependency runs before its dependant. */
function topologicalOrder(sources: Source[]): Source[] {
  const sourceMap = new Map<string, Source>(sources.map((s) => [s.id, s]));
  const result:   Source[] = [];
  const visited   = new Set<string>();
  const visiting  = new Set<string>();

  function visit(id: string): void {
    if (visited.has(id)) return;
    if (visiting.has(id)) throw new Error(`[Runner] Circular dependency at '${id}'`);
    visiting.add(id);
    for (const dep of sourceMap.get(id)?.depends_on ?? []) visit(dep);
    visiting.delete(id);
    visited.add(id);
    if (sourceMap.has(id)) result.push(sourceMap.get(id)!);
  }

  for (const s of sources) visit(s.id);
  return result;
}

// ─── Single-source executor ───────────────────────────────────────────────────

export async function runSource(source: Source): Promise<SourceResult> {
  const sid       = source.id;
  const sheetId   = source.sheet_id
  const sheetName = source.sheet_name;

  // E — Extract
  console.log(`[Runner:${sid}] E — extracting from sheet '${sheetName}'`);
  const rawData = await fetchSheetData(sheetId, sheetName);
  const records = parseOjectFromSheet(rawData);
  if (!records.length) throw new Error(`Sheet '${sheetName}' returned no data.`);

  // t — Stage raw (append + deduplicate)
  const rawBlob = source.raw_blob;
  console.log(`[Runner:${sid}] t — staging raw → ${rawBlob}`);
  const rawCsv = await parseCsvFromObject(records);
  await appendCsvToGCS(rawCsv, rawBlob);
  const gcsRaw = `gs://${process.env.GCS_BUCKET_NAME}/${rawBlob}`;

  let gcsProcessed: string | undefined;

  // T — Transform (cleaner, if enabled)
  if (source.cleaner?.enabled && source.cleaner.path) {
    const cleanerPath = source.cleaner.path;
    console.log(`[Runner:${sid}] T — cleaning with '${cleanerPath}'`);
    const rawBuffer = await downloadCsvFromGCS(rawBlob);
    const rawRows   = Papa.parse<Record<string, any>>(rawBuffer.toString(), {
      header:         true,
      skipEmptyLines: true,
    }).data;

    const cleaner    = importModule(cleanerPath);
    const cleanedRows: Record<string, any>[] = cleaner.clean(rawRows);
    if (!cleanedRows?.length) {
      throw new Error(`Cleaner '${cleanerPath}' produced no rows for source '${sid}'.`);
    }

    const processedBlob = source.processed_blob!;
    console.log(`[Runner:${sid}] T — staging processed → ${processedBlob}`);
    await uploadCsvToGCS(Papa.unparse(cleanedRows), processedBlob);
    gcsProcessed = `gs://${process.env.GCS_BUCKET_NAME}/${processedBlob}`;
  } 
  else if (source.processed_blob) {
    await uploadCsvToGCS(rawCsv, source.processed_blob);
    gcsProcessed = `gs://${process.env.GCS_BUCKET_NAME}/${source.processed_blob}`;
  }

  // L — Load to MySQL
  let rowsLoaded = 0;
  if (source.loader?.enabled && source.loader.path) {
    const loaderPath = source.loader.path;
    console.log(`[Runner:${sid}] L — loading with '${loaderPath}'`);
    const loader = importModule(loaderPath);
    rowsLoaded   = await loader.load();
  }

  const result: SourceResult = {
    status:        "ok",
    rows_in_sheet: records.length,
    gcs_raw:       gcsRaw,
    rows_loaded:   rowsLoaded,
  };
  if (gcsProcessed) result.gcs_processed = gcsProcessed;

  console.log(`[Runner:${sid}] done — rows_loaded=${rowsLoaded}`);
  return result;
}

// ─── Pipeline entry point ─────────────────────────────────────────────────────

export async function runPipeline(
  sourceIds?: string[]
): Promise<Record<string, SourceResult>> {
  const manifest = JSON.parse(fs.readFileSync(MANIFEST_PATH, "utf-8"));

  let sources: Source[] = manifest.sources.filter((s: Source) => s.enabled !== false);
  if (sourceIds?.length) {
    sources = sources.filter((s) => sourceIds.includes(s.id));
  }

  const ordered = topologicalOrder(sources);
  const results: Record<string, SourceResult> = {};

  for (const source of ordered) {
    try {
      results[source.id] = await runSource(source);
    } catch (err: any) {
      results[source.id] = { status: "error", error: err.message };
      console.error(`[Runner:${source.id}] ERROR — ${err.message}`);
    }
  }

  return results;
}

File: ingestion/python_api/engine/runner.py
# =============================================================================
# File: engine/runner.py
# Purpose: Generic EtLT engine.  Reads pipeline_manifest.json, resolves the
#          dependency order, and executes the Extract → stage-raw → Transform
#          → stage-processed → Load flow for each enabled source.
#
#   Each source entry in the manifest declares:
#     sheet_id / sheet_name  – where to extract data from
#     raw_blob               – GCS path for the raw staged CSV
#     processed_blob         – GCS path for the cleaned CSV (null if no cleaner)
#     cleaner.enabled        – whether to run a transform step
#     cleaner.path           – dotted module path, e.g. "EtLT.transform.expenses"
#                              module must expose:  clean(df) -> df
#     loader.enabled         – whether to load into MySQL
#     loader.path            – dotted module path, e.g. "EtLT.load.transactions"
#                              module must expose:  load() -> int
#     depends_on             – list of source ids that must run first
# =============================================================================

import importlib
import json
import os
import sys
from graphlib import TopologicalSorter
from io import BytesIO
from pathlib import Path

import pandas as pd

from EtLT.staging.gcs_service import append_to_gcs, download_from_gcs, upload_to_gcs
from EtLT.extract.spreadsheet_service import fetch_sheet_data, dataframe_to_csv_bytes

_MANIFEST_PATH = Path(os.getenv("MANIFEST_PATH", str(Path(__file__).parent.parent.parent / "pipeline_manifest.json")))


# ─── Helpers ──────────────────────────────────────────────────────────────────


def _import_module(dotted_path: str):
    """Import a module by dotted path (e.g. 'transform.expenses')."""
    return importlib.import_module(dotted_path)


def _topological_order(sources: list) -> list:
    """Return sources sorted so that every dependency runs before its dependant."""
    graph = {s["id"]: set(s.get("depends_on", [])) for s in sources}
    ts = TopologicalSorter(graph)
    order = list(ts.static_order())
    source_map = {s["id"]: s for s in sources}
    return [source_map[sid] for sid in order if sid in source_map]


# ─── Single-source executor ───────────────────────────────────────────────────

def run_source(source: dict) -> dict:
    """
    Execute the full EtLT flow for one manifest entry.
    Returns a result dict with status, row counts, and GCS URIs.
    """
    sid = source["id"]
    creds = os.getenv("GOOGLE_SHEETS_CREDENTIALS") or os.getenv("GOOGLE_SHEETS_CREDENTIALS_LOCAL")
    sheet_id = source["sheet_id"]
    sheet_name = source["sheet_name"]

    # Extract
    print(f"[runner:{sid}] E — extracting from sheet '{sheet_name}'")
    df = fetch_sheet_data(sheet_id, sheet_name, creds, as_type="dataframe")
    if df is None or df.empty:
        raise ValueError(f"Sheet '{sheet_name}' returned no data.")

    # Stage raw (append + deduplicate)
    raw_blob = source["raw_blob"]
    print(f"[runner:{sid}] t — staging raw → {raw_blob}")
    gcs_raw_uri = append_to_gcs(new_df=df, blob_name=raw_blob)

    gcs_processed_uri = None

    # Transform (cleaner, if enabled)
    cleaner_cfg = source.get("cleaner", {})
    if cleaner_cfg.get("enabled"):
        cleaner_path = cleaner_cfg["path"]
        print(f"[runner:{sid}] T — cleaning with '{cleaner_path}'")
        raw_bytes = download_from_gcs(raw_blob)
        raw_df = pd.read_csv(BytesIO(raw_bytes), encoding="utf-8-sig")
        cleaner = _import_module(cleaner_path)
        cleaned_df = cleaner.clean(raw_df)
        if cleaned_df is None or cleaned_df.empty:
            raise ValueError(
                f"Cleaner '{cleaner_path}' produced an empty DataFrame for source '{sid}'."
            )
        processed_blob = source["processed_blob"]
        print(f"[runner:{sid}] T — staging processed → {processed_blob}")
        gcs_processed_uri = upload_to_gcs(
            file_bytes=dataframe_to_csv_bytes(cleaned_df),
            destination_blob_name=processed_blob,
        )
    elif source.get("processed_blob"):
        # If no cleaner but processed_blob is defined, copy raw to processed for downstream steps
        processed_blob = source["processed_blob"]
        print(f"[runner:{sid}] T — no cleaner, copying raw → {processed_blob}")
        raw_bytes = download_from_gcs(raw_blob)
        gcs_processed_uri = upload_to_gcs(
            file_bytes=raw_bytes,
            destination_blob_name=processed_blob,
        )

    # Load to MySQL 
    loader_cfg  = source.get("loader", {})
    rows_loaded = 0
    if loader_cfg.get("enabled"):
        loader_path = loader_cfg["path"]
        print(f"[runner:{sid}] L — loading with '{loader_path}'")
        loader = _import_module(loader_path)
        rows_loaded = loader.load()

    result = {
        "status": "ok",
        "rows_in_sheet": len(df),
        "gcs_raw": gcs_raw_uri,
        "rows_loaded": rows_loaded,
    }
    if gcs_processed_uri:
        result["gcs_processed"] = gcs_processed_uri

    print(f"[runner:{sid}] done — rows_loaded={rows_loaded}")
    return result


# ─── Pipeline entry point ─────────────────────────────────────────────────────

def run_pipeline(source_ids: list = None) -> dict:
    """
    Run the EtLT pipeline for all enabled sources (or a filtered subset).

    Execution order respects each source's depends_on list.
    Failures are isolated: one failed source does not abort the rest.

    Args:
        source_ids: Optional list of source IDs to run.  If None, all
                    enabled sources in the manifest are executed.

    Returns:
        Dict mapping source_id → result dict with keys:
          status ("ok" | "error"), rows_in_sheet, gcs_raw,
          rows_loaded, gcs_processed (if applicable), error (if failed).
    """
    with open(_MANIFEST_PATH, encoding="utf-8") as f:
        manifest = json.load(f)

    sources = [s for s in manifest["sources"] if s.get("enabled", True)]
    if source_ids:
        sources = [s for s in sources if s["id"] in source_ids]

    ordered = _topological_order(sources)
    results: dict = {}

    for source in ordered:
        sid = source["id"]
        try:
            results[sid] = run_source(source)
        except Exception as exc:
            results[sid] = {"status": "error", "error": str(exc)}
            print(f"[runner:{sid}] ERROR — {exc}", file=sys.stderr)

    return results
